### DISCUSSION TITLE
Docker Desktop setup for dummies?

### DISCUSSION THREAD

**Discussion by Benjin** [qdm12/ddns-updater#792] [Q&A]
I've installed Docker Desktop on my machine (Windows host), and downloaded/run the ddns-updater image, and now it's time to follow the [Container instructions](https://github.com/qdm12/ddns-updater?tab=readme-ov-file#container).

But I'm not really sure where to start there.  It says to "create a directory", but as a Docker newbie, it's not obvious to me where or how to create said directory.  Is there a way to open a shell within the docker container in order to run those commands?

From a usability standpoint, I was expecting to:
1. Start the unconfigured docker container
2. Go to localhost:8000
3. Use the web UI to add my JSON config

but nothing seems to be running on localhost:8000.  Notably, there's no local IP listed for the http server, like there is for the healthcheck server.  That doesn't seem right to me:
![image](https://github.com/user-attachments/assets/a574d09d-81a8-4db8-bd0f-424c45cb2801)

**Comment by qdm12** [comment]
> t's not obvious to me where or how to create said directory. Is there a way to open a shell within the docker container in order to run those commands?

Create it on the host Machine, so likely through a termimal (over ssh etc.). And then it will be used by the contaimer through the bind mount `-v` flag 😉

Edit: you can also use the CONFIG environment variable and stuff the json on a single line there, if you prefer.


> From a usability standpoint, I was expecting to

Definitely something to be added in the future (not actively being worked on yet though), even without a json but with nice buttons etc. But for now the code internals are not even ready to live patch settings at runtime. Anyway, today the only way is to define settings at container start.

> Notably, there's no local IP listed for the http server, like there is for the healthcheck server. That doesn't seem right to me:

`[::]` means the http server listens on all interfaces, ipv4+ipv6 (unlike `0.0.0.0` which is all interfaces for ipv4 only). If you can't access the web ui, did you publish the port 8000 correctly?

For the healthcheck server, it listens on `127.0.0.1` so it can only be accessed by the container itself (localhost from its point of view), and can't be accessed even if you publish port 9999. The reason is this is for the docker healthcheck - running within the container only - and so no need to expose it and increase the "surface of attack" security-wise.

**Reply by Benjin** [reply]
> Create it on the host machine

Huh, not at all what I expected the solution to be.  In any case, I found that Docker Desktop has a way to browse and edit existing files in the container, and that `updater/data/config.json` was automatically created at first run, so I was able to do everything through that!

> Did you publish port 8000 correctly?

I'm guessing not - also something that I expected to happen automatically from my experience with running other docker images.  For example, 13ft's docker image exposes their port automatically.  Any reason ddns-updater doesn't do the same thing?

![image](https://github.com/user-attachments/assets/b86426e1-a5c6-40c9-a9bd-af69de861e8d)

**Reply by qdm12** [reply]
> Docker Desktop

Sorry I didn't see this/read too quickly; I suppose it runs using WSL so you can in your terminal (cmd, powershell etc.) run `wsl` to run the shell commands, and then configure the bind mount in the docker desktop UI (or directly run the docker run command in that wsl shell).

> was automatically created at first run, so I was able to do everything through that!

There is no live reload feature in ddns-updater for now, so I suppose you this won't change the program state, and won't be persisted across container restarts. Following the steps above should do the trick though.



> I'm guessing not - also something that I expected to happen automatically from my experience with running other docker images.

As far as I know, you need to specify the ports published yourself, there is no automation. We can describe what ports are "exposed" (documentation purpose really) in the docker image, but that's it. You would need to map the port `8000:8000` for example (or any-value:8000),similarly to your other container with mapping 5000:5000

**Reply by Benjin** [reply]
Got it working after slogging over to the commandline; thanks for the patience!  Might be nice to update the README with those basic instructions (`docker run -d -p 8000:8000 qmcgaw/ddns-updater:latest` and that you can edit the config through Docker Desktop if you don't want to mount a volume).  Or I can send a PR.

[13ft](https://github.com/wasi-master/13ft/blob/main/docker-compose.yaml) is a simple container that automatically published the ports, I guess their secret sauce was the docker-compose.yaml file.

**Reply by qdm12** [reply]
13ft runs as root by default (unlike ddns-updater) and does not have any bind mount, that's why there isn't additional setup steps. Note there is also a docker-compose.yml for ddns-updater: https://github.com/qdm12/ddns-updater/blob/master/docker-compose.yml

The docker run command is already described in step 4 of https://github.com/qdm12/ddns-updater?tab=readme-ov-file#container

> and that you can edit the config through Docker Desktop if you don't want to mount a volume

So you would create an anonymous volume and edit the file through the ui right? Could you share screenshots? Does this persist across container restarts? Thanks!
